clean up realtime AD cache on old coordinating node

ylwu-amzn opened the pull request:
Signed-off-by: Yaliang Wu <ylwu@amazon.com>

### Description
Init realtime task cache and clean up realtime task cache on old coordinating node. Realtime AD depends on job scheduler to choose node (job coordinating node) to run AD job. Nodes have primary or replica shard of AD job index are candidate to run AD job. Job scheduler will build hash ring on these candidate nodes and choose one to run AD job. If AD job index shard relocated, for example new node added into cluster, then job scheduler will rebuild hash ring and may choose different node to run AD job. So we need to init realtime task cache on new AD job coordinating node and clean up cache on old coordinating node.

Track last AD realtime job run time and cleanup realtime cache in hourly cron if last realtime job run was 2 detector interval ago.
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ylwu-amzn commented on src/main/java/org/opensearch/ad/transport/AnomalyResultTransportAction.java:
```diff
@@ -437,110 +442,140 @@ private void handleException(Exception e) {
+        }, e -> LOG.error("Failed to init realtime task cache for " + detectorId, e));
+    }
+
+    private void executeAnomalyDetection(
```
Just wrap old logic in this method. The old realtime job run logic still the same.

ohltyler commented on src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorJobActionHandler.java:
```diff
@@ -333,14 +337,18 @@ public void stopAnomalyDetectorJob(String detectorId) {
                     logger.info("AD model deleted successfully for detector {}", detectorId);
-                    adTaskManager.updateLatestRealtimeTask(detectorId, ADTaskState.STOPPED, null, listener);
+                    // StopDetectorTransportAction will send out DeleteModelAction which will clear all realtime cache.
+                    // Pass null transport service to method "stopLatestRealtimeTask" to not re-clear coordinating node cache.
+                    adTaskManager.stopLatestRealtimeTask(detectorId, ADTaskState.STOPPED, null, null, listener);
```
Why not clear here? Can you point me to the line where the cache is initially cleared for the coordinating node?

ylwu-amzn replied:
Check line 316, when stop detector, client will send out `StopDetectorAction` , and `StopDetectorTransportAction` will send out `DeleteModelAction` which is broadcast request;  then `DeleteModelTransportAction` will clean up realtime caches for the detector.

ylwu-amzn commented:
@kaituo , please check if HC realtime AD cache cleaned up when coordinating node change, I haven't checked that part yet.

ylwu-amzn commented on src/main/java/org/opensearch/ad/NodeStateManager.java:
```diff
@@ -286,12 +286,21 @@ public void addPressure(String nodeId, String detectorId) {
     public void resetBackpressureCounter(String nodeId, String detectorId) {
         Map<String, BackPressureRouting> routingMap = backpressureMuter.get(detectorId);
         if (routingMap == null || routingMap.isEmpty()) {
+            // TODO: confirm with Kaituo, do we need to remove detector's backpressure counter for all nodes/
```
@kaituo , can you confirm if it's ok to remove detector's backpressure counter for all nodes here? Or we just need to remove counter for the passed in node?

ylwu-amzn commented on src/main/java/org/opensearch/ad/ml/ModelManager.java:
```diff
@@ -475,6 +475,8 @@ public void stopModel(String detectorId, String modelId, ActionListener<Void> li
     /**
      * Permanently deletes models hosted in memory and persisted in index.
      *
+     * TODO: confirm with Kaituo, when will we call this clear method, only see when call this method stop realtime job
```
@kaituo , I only see this method called in `DeleteModelTransportAction` , which will be executed when stop detector. If detector still running, just AD job coordinating node changed, then these models will not be cleared. 

Should we clean up all models in cache for detector if its AD job coordinating node changed?

ylwu-amzn replied:
Discussed with Kaituo offline, if node added/removed, we can check if model owning node can match local node or not. If can't match, we can delete model cache

ylwu-amzn commented on src/main/java/org/opensearch/ad/rest/handler/IndexAnomalyDetectorJobActionHandler.java:
```diff
@@ -293,7 +297,7 @@ public void stopAnomalyDetectorJob(String detectorId) {
                     AnomalyDetectorJob job = AnomalyDetectorJob.parse(parser);
                     if (!job.isEnabled()) {
-                        adTaskManager.updateLatestRealtimeTask(detectorId, ADTaskState.STOPPED, null, listener);
+                        adTaskManager.stopLatestRealtimeTask(detectorId, ADTaskState.STOPPED, null, transportService, listener);
```
Here, just rename method `updateLatestRealtimeTask` as `stopLatestRealtimeTask`

ylwu-amzn commented on src/main/java/org/opensearch/ad/transport/ForwardADTaskTransportAction.java:
```diff
@@ -125,6 +151,12 @@ protected void doExecute(Task task, ForwardADTaskRequest request, ActionListener
+                    // change, then we should clean up cache on old coordinating node.
+                    stateManager.removeBackpressureCounter(detectorId);
+                    featureManager.clear(detectorId);
+                    entityColdStarter.removeDoorKeeper(detectorId);
```
Entity cold starter runs on work nodes, hourly cron will clean up expired cache. Will not clean up door keeper here.

ylwu-amzn commented on src/main/java/org/opensearch/ad/transport/ForwardADTaskTransportAction.java:
```diff
@@ -125,6 +151,12 @@ protected void doExecute(Task task, ForwardADTaskRequest request, ActionListener
+                    stateManager.removeBackpressureCounter(detectorId);
+                    featureManager.clear(detectorId);
+                    entityColdStarter.removeDoorKeeper(detectorId);
+                    priorityCache.get().removeDetectorCache(detectorId);
```
Priority cache will be cleaned up in hourly cron. Ideally we can use similar way of single entity detector to clean up model cache when receive node added/removed event. Will create an issue to track this.

ylwu-amzn replied:
https://github.com/opensearch-project/anomaly-detection/issues/210

kaituo commented on src/main/java/org/opensearch/ad/transport/DeleteModelTransportAction.java:
```diff
@@ -65,7 +66,8 @@ public DeleteModelTransportAction(
         CacheProvider cache,
-        ADTaskCacheManager adTaskCacheManager
+        ADTaskCacheManager adTaskCacheManager,
+        NodeStateManager stateManager
```
We have had a NodeStateManager. The name transportStateManager is outdated though.

kaituo commented on src/main/java/org/opensearch/ad/transport/ForwardADTaskTransportAction.java:
```diff
@@ -125,6 +141,10 @@ protected void doExecute(Task task, ForwardADTaskRequest request, ActionListener
+                    // If hash ring changed like new node added when scale out, the realtime job coordinating node may
+                    // change, then we should clean up cache on old coordinating node.
+                    stateManager.removeBackpressureCounter(detectorId);
+                    featureManager.clear(detectorId);
```
Adding stateManager.clear(detectorId) as well?

kaituo commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -1953,117 +1965,85 @@ public void updateLatestRealtimeTask(
         }
-        Float newInitProgress = initProgress;
-        String newError = error;
+        Float finalInitProgress = initProgress;
+        String finalError = error;
+        String finalNewState = newState;
```
why do you need to do these assignments?

ylwu-amzn replied:
Variable used in lambda expression should be final or effectively final

kaituo replied:
Can you add a comment for this?

ylwu-amzn replied:
Sure

kaituo commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -2968,4 +2878,111 @@ public int getLocalAdAssignedBatchTaskSlot() {
+     * @param transportService transport service
+     * @param size return how many tasks
+     */
+    public void maintainRunningHistoricalTasks(TransportService transportService, int size) {
```
Do you think it is necessary to add maintenance for memory tracker just in case you don't release memory? I have one in PriorityCache for HC detector: https://github.com/opensearch-project/anomaly-detection/blob/main/src/main/java/org/opensearch/ad/caching/PriorityCache.java#L517-L539

ylwu-amzn replied:
For historical analysis, we will clean up cache if task finished normally or exception happens. If we fail to remove cache, the reset logic will try to clean up again, and hourly cron will clean up as well. So I think it's safe. But I'm open to add more check to make sure no memory leak. Will create an issue to track this.

kaituo replied:
Sounds good to me.

ylwu-amzn commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -1446,7 +1445,7 @@ private boolean validateDetector(AnomalyDetector detector, ActionListener<Anomal
     }
 
-    public void executeAnomalyDetector(
+    private void executeAnomalyDetector(
```
Can you suggest a new name?

kaituo replied:
Yes, it is clearer

ylwu-amzn commented on src/main/java/org/opensearch/ad/transport/DeleteModelTransportAction.java:
```diff
@@ -65,7 +66,8 @@ public DeleteModelTransportAction(
         CacheProvider cache,
-        ADTaskCacheManager adTaskCacheManager
+        ADTaskCacheManager adTaskCacheManager,
+        NodeStateManager stateManager
```
will rename the variable and remove stateManager

ylwu-amzn commented on src/main/java/org/opensearch/ad/transport/ForwardADTaskTransportAction.java:
```diff
@@ -125,6 +141,10 @@ protected void doExecute(Task task, ForwardADTaskRequest request, ActionListener
+                    // If hash ring changed like new node added when scale out, the realtime job coordinating node may
+                    // change, then we should clean up cache on old coordinating node.
+                    stateManager.removeBackpressureCounter(detectorId);
+                    featureManager.clear(detectorId);
```
sure, will add

ylwu-amzn commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -1482,11 +1484,13 @@ private void createNewADTask(
         String userName = user == null ? null : user.getName();
         Instant now = Instant.now();
         String taskType = getADTaskType(detector, detectionDateRange).name();
+        String coordinatingNode = detectionDateRange == null ? realtimeADCoordinatingNode : clusterService.localNode().getId();
```
Check `ADTaskManager#startHistoricalAnalysis` line 407, it will choose owning node and forward START action to coordinating node. Then check `ForwardADTaskTransportAction` line 125, it starts historical analysis on coordinating node.

kaituo replied:
I see.  Is it more clearer that you repurpose realtimeADCoordinatingNode as coordinatingNode and let the caller decide what node to pass in? Currently, the precondition is that the caller of this method is from a coordinating node.  It is easy for others to forget and make mistakes.

ylwu-amzn replied:
Sure, will tune this part.

kaituo commented on src/main/java/org/opensearch/ad/cluster/ADClusterEventListener.java:
```diff
@@ -128,30 +128,26 @@ public void clusterChanged(ClusterChangedEvent event) {
-                                }
+                        String localNodeId = event.state().nodes().getLocalNode().getId();
+                        Set<String> modelIds = modelManager.getAllModelIds();
+                        for (String modelId : modelIds) {
```
Previously, we do it whenever hash ring rebuild successfully (depending on cool down is past and other concurrent conditions).  Now, you would build the hash ring without the cooldown.  So hasRingBuildDone only means historical analysis's hash ring change, but real time does not necessarily so.  So real time analysis might be use the old hash ring while we clear models here.  Would you consider moving the logic to HashRing when the real time hash ring actually changes?

ylwu-amzn replied:
The cluster node added/removed event is limited, generally it only happens when cluster starts or scales. It may have some extra checks during the cool down period, but that impact should be minor. How about add some TODO and fix it later?

kaituo commented:
Changes look good.  Wait for you to fix test failures:

 1073 tests completed, 2 failed, 8 skipped
Tests with failures:
 - org.opensearch.ad.cluster.ADClusterEventListenerTests.testInprogress
 - org.opensearch.ad.cluster.ADClusterEventListenerTests.testNotRecovered

ylwu-amzn commented:
```
2021-09-10T17:29:26.0307465Z Tests with failures:
2021-09-10T17:29:26.0309610Z  - org.opensearch.ad.transport.MultiEntityResultTests.testNonEmptyFeatures
2021-09-10T17:29:26.0313717Z  - org.opensearch.ad.transport.AnomalyDetectorJobTransportActionTests.testRunMultipleTasksForHistoricalAnalysis
2021-09-10T17:29:26.0318592Z  - org.opensearch.ad.transport.DeleteAnomalyResultsTransportActionTests.testDeleteADResultAction
```

> Changes look good. Wait for you to fix test failures:
> 
> 1073 tests completed, 2 failed, 8 skipped
> Tests with failures:
> 
> * org.opensearch.ad.cluster.ADClusterEventListenerTests.testInprogress
> * org.opensearch.ad.cluster.ADClusterEventListenerTests.testNotRecovered

Thanks @kaituo , I have fixed failed test and commented out flaky test case. But link check failed due to wrong link, which fixed in your PR #202

ylwu-amzn commented:
Flaky tests
```
1092 tests completed, 1 failed, 9 skipped
Tests with failures:
 - org.opensearch.ad.ratelimit.CheckpointWriteWorkerTests.testTriggerAutoFlush
```

kaituo commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -196,6 +197,8 @@
 public class ADTaskManager {
     public static final String AD_TASK_LEAD_NODE_MODEL_ID = "ad_task_lead_node_model_id";
     public static final String AD_TASK_MAINTAINENCE_NODE_MODEL_ID = "ad_task_maintainence_node_model_id";
+    // HC batch task timeout after 10 minutes
```
Is this per task or per detector (i.e., total processing time of the 1K entity tasks should be less than 10 mins)?  According to ADTaskManager.resetHistoricalDetectorTaskState, this sounds like the latter. If so, can we assume all historical analyses finishes within 10 minutes?

kaituo commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -818,7 +821,7 @@ public void stopDetector(
                     (task) -> stopHistoricalAnalysis(detectorId, task, user, listener),
                     transportService,
-                    true,
+                    false,// don't need to reset task state when stop detector
```
why we don't need to reset any more?

kaituo commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -2371,9 +2407,13 @@ public void updateADHCDetectorTask(
                 logger.error("Failed to update detector task " + taskId, e);
                 adTaskCacheManager.releaseTaskUpdatingSemaphore(detectorId);
             }
+        } else if (!adTaskCacheManager.isHCTaskCoordinatingNode(detectorId)) {
```
When would we throw exception? Did you return listener in the block 2406~2409?

kaituo commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -198,6 +197,8 @@
     public static final String AD_TASK_LEAD_NODE_MODEL_ID = "ad_task_lead_node_model_id";
     public static final String AD_TASK_MAINTAINENCE_NODE_MODEL_ID = "ad_task_maintainence_node_model_id";
+    // HC batch task timeout after 10 minutes
+    public static final int HC_BATCH_TASK_CACHE_TIMEOUT_IN_MILLIS = 600_000;
```
dataMigrator and searchFeatureDao are unused.

kaituo commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -1135,16 +1105,30 @@ public void stopDetector(
+                    && detector.isMultientityDetector()
+                    && !isNullOrEmpty(taskProfile.getEntityTaskProfiles())) {
+                    // If coordinating node restarted, HC detector cache on it will be gone. But worker node still
+                    // runs entity tasks, we'd better stop these entity tasks to clean up resource ealier.
```
ealier -> earlier

ylwu-amzn commented on src/test/java/org/opensearch/ad/transport/AnomalyDetectorJobTransportActionTests.java:
```diff
@@ -229,11 +230,16 @@ public void testStartHistoricalAnalysisForMultiCategoryHCWithUser() throws IOExc
             waitUntil(() -> {
                 try {
                     ADTask task = getADTask(taskId);
+                    if (ADTaskState.RUNNING.name().equals(task.getState())) {
+                        ProfileRequest profileRequest = new ProfileRequest(detectorId, ImmutableSet.of(DetectorProfileName.AD_TASK), true);
+                        ProfileResponse profileResponse = client().execute(ProfileAction.INSTANCE, profileRequest).actionGet(10_000);
```
Oh, forgot to remove these code, just some debugging code

ylwu-amzn commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -604,54 +603,19 @@ public void checkTaskSlots(
-                        1
-                    );
-                }
+                int approvedTaskSlots = detector.isMultientityDetector()
```
Sure, will add more comments

kaituo commented on src/main/java/org/opensearch/ad/task/ADBatchTaskRunner.java:
```diff
@@ -1270,21 +1279,28 @@ private float calculateInitProgress(String taskId) {
+            // clean up pending and running entity on coordinating node
+            adTaskCacheManager.clearPendingEntities(detectorId);
+            adTaskCacheManager.removeRunningEntity(detectorId, adTaskManager.convertEntityToString(adTask));
+            throw new ADTaskCancelledException(
```
Thanks for your explanation.  

For the code path I asked, want to double check if exceptions would be caught by your code:

1)ADBatchAnomalyResultTransportAction.doExecute does not catch exceptions so the exception will be thrown to the transport layer?
ADBatchTaskRunner.checkIfADTaskCancelledAndCleanupCache <- ADBatchTaskRunner.forwardOrExecuteADTask <- ADBatchTaskRunner.run <- ADBatchAnomalyResultTransportAction.doExecute

2)whom would handle the exception from the scheduled run of startNewEntityTaskLane?  
ADBatchTaskRunner.checkIfADTaskCancelledAndCleanupCache <- ADBatchTaskRunner.forwardOrExecuteADTask <- ADBatchTaskRunner.startNewEntityTaskLane <- ADBatchTaskRunner.workerNodeResponseListener (
threadPool
                        .schedule(
                            () -> startNewEntityTaskLane(adTask, transportService),
                            TimeValue.timeValueSeconds(SLEEP_TIME_FOR_NEXT_ENTITY_TASK_IN_MILLIS),
                            AD_BATCH_TASK_THREAD_POOL_NAME
                        );
)

ylwu-amzn replied:
[TransportAction](https://github.com/opensearch-project/OpenSearch/blob/0ba0e7cc26060f964fcbf6ee45bae53b3a9941d0/server/src/main/java/org/opensearch/action/support/TransportAction.java#L192) will catch exception and call onFailure method. Changed to catch all exceptions in `forwardOrExecuteADTask`

kaituo commented:
Could you provide context for your most recent commits?

https://github.com/ylwu-amzn/anomaly-detection-2/pull/209/commits/4cdd2e3155ad9947347bda9786989411eef35107

https://github.com/ylwu-amzn/anomaly-detection-2/pull/209/commits/2cca97b0ce4d3bd9bd142353477793e766cc8016

ylwu-amzn commented:
> Could you provide context for your most recent commits?
> 
> [ylwu-amzn@4cdd2e3](https://github.com/ylwu-amzn/anomaly-detection-2/commit/4cdd2e3155ad9947347bda9786989411eef35107)

The entity filter doesn't work for multi-category HC, fixed that for multi-category HC.

Found some issue when many entities running concurrently, may show more running entities than actual tasks on worker node in profile API. Fixed the task cache to make the running entities correct. 

We supported stopping HC historical analysis after task running on worker node. For multi-category HC, the stress testing shows it may take a long time(more than 1 minute for 1.8 billions of doc and 4k entities). We should not let user wait for a long time until querying top entities done and task running if they want to stop historical analysis. This commit tuned HC task run state to support task level stopped status, so it's safe for user to stop historical analysis at any time. 

> [ylwu-amzn@2cca97b](https://github.com/ylwu-amzn/anomaly-detection-2/commit/2cca97b0ce4d3bd9bd142353477793e766cc8016)

Make the entity movement between pending, temp and running queues consistent: one entity can only exist in one queue at the same time.

1. Poll entity: move entity from pending queue to temp queue
2. If failed to send entity task to worker node like no available task slots: move entity from temp queue to pending queue
If send entity task to worker node successfully: move entity from temp queue to running queue
3. Entity task done(FINISHED/FAILED/STOPPED) on worker node: remove entity from running queue

ohltyler commented on src/main/java/org/opensearch/ad/task/ADHCBatchTaskCache.java:
```diff
@@ -220,7 +218,8 @@ public boolean hasRunningEntity() {
 
     public boolean removeRunningEntity(String entity) {
-        return this.runningEntities.remove(entity);
+        // entity may still in temp queue, so need to remove from both temp and running queues
```
Can you explain how the entity could be in either temp or running queue? Doesn't `moveToRunningEntity` remove such entity from the temp queue on line 170?

ylwu-amzn replied:
In normal case, the entity will be moved to running queue if entity task dispatched to worker node successfully. If failed to dispatch to worker node, it will still stay in temp queue, check `ADBatchTaskRunner#workerNodeResponseListener`. Then will send entity task done message to coordinating node to move to pending queue if exception is retryable or remove entity from cache if not retryable.

kaituo replied:
Could you add these comments?  They are really helpful.

Also, for HC detector, you will dispatch both detector level and task level tasks to worker nodes?  If yes, what's purpose of dispatching a detector level task to a worker node?  Could you point me any code link?

ylwu-amzn replied:
For HC detector, will not dispatch detector level task to worker node, only dispatch entity level task. Check `ADBatchTaskRunner#forwardOrExecuteADTask`,  line 553 creates new entity task , `ADTask adEntityTask = new ADTask.Builder()`, line 578 forward entity task to worker node, `forwardOrExecuteEntityTask(adEntityTask, transportService, workerNodeResponseListener);`

ohltyler commented on src/main/java/org/opensearch/ad/task/ADHCBatchTaskRunState.java:
```diff
@@ -11,19 +11,25 @@
 public class ADHCBatchTaskRunState {
 
+    // HC batch task run state will expire after 60 seconds after last task run time or task cancelled time.
+    public static final int HC_TASk_RUN_STATE_TIMEOUT_IN_MILLIS = 60_000;
```
nit: `TASk` -> `TASK`

ylwu-amzn replied:
sure, will fix

kaituo commented:
https://github.com/opensearch-project/anomaly-detection/pull/209#discussion_r714353576

Cannot comment on the previous conversation.  Got the question about dispatching detector level task while reading your comment in workerNodeResponseListener.  is it outdated as you said we won't dispatch detector level tasks to worker node?

```
if (adTask.getDetector().isMultientityDetector()) {
                // For HC detector, the first task is not entity task, it's HC detector task. But no matter it's entity
                // or detector level task, we should always send back entity task done message for HC detector.
                adTaskManager.entityTaskDone(adTask, e, transportService);
                if (adTaskCacheManager.getAvailableNewEntityTaskLanes(adTask.getDetectorId()) > 0) {
                    // When reach this line, it means entity task failed to start on worker node
                    // Sleep some time before starting new task lane.
                    threadPool
                        .schedule(
                            () -> startNewEntityTaskLane(adTask, transportService),
                            TimeValue.timeValueSeconds(SLEEP_TIME_FOR_NEXT_ENTITY_TASK_IN_MILLIS),
                            AD_BATCH_TASK_THREAD_POOL_NAME
                        );
                }
            }
```

ylwu-amzn commented:
> [#209 (comment)](https://github.com/opensearch-project/anomaly-detection/pull/209#discussion_r714353576)
> 
> Cannot comment on the previous conversation. Got the question about dispatching detector level task while reading your comment in workerNodeResponseListener. is it outdated as you said we won't dispatch detector level tasks to worker node?
> 
> ```
> if (adTask.getDetector().isMultientityDetector()) {
>                 // For HC detector, the first task is not entity task, it's HC detector task. But no matter it's entity
>                 // or detector level task, we should always send back entity task done message for HC detector.
>                 adTaskManager.entityTaskDone(adTask, e, transportService);
>                 if (adTaskCacheManager.getAvailableNewEntityTaskLanes(adTask.getDetectorId()) > 0) {
>                     // When reach this line, it means entity task failed to start on worker node
>                     // Sleep some time before starting new task lane.
>                     threadPool
>                         .schedule(
>                             () -> startNewEntityTaskLane(adTask, transportService),
>                             TimeValue.timeValueSeconds(SLEEP_TIME_FOR_NEXT_ENTITY_TASK_IN_MILLIS),
>                             AD_BATCH_TASK_THREAD_POOL_NAME
>                         );
>                 }
>             }
> ```

Good catch, will fix the comment. Let me know when you finish reviewing all files. Will fix all comments in a new commit.

kaituo commented:
> > [#209 (comment)](https://github.com/opensearch-project/anomaly-detection/pull/209#discussion_r714353576)
> > Cannot comment on the previous conversation. Got the question about dispatching detector level task while reading your comment in workerNodeResponseListener. is it outdated as you said we won't dispatch detector level tasks to worker node?
> > ```
> > if (adTask.getDetector().isMultientityDetector()) {
> >                 // For HC detector, the first task is not entity task, it's HC detector task. But no matter it's entity
> >                 // or detector level task, we should always send back entity task done message for HC detector.
> >                 adTaskManager.entityTaskDone(adTask, e, transportService);
> >                 if (adTaskCacheManager.getAvailableNewEntityTaskLanes(adTask.getDetectorId()) > 0) {
> >                     // When reach this line, it means entity task failed to start on worker node
> >                     // Sleep some time before starting new task lane.
> >                     threadPool
> >                         .schedule(
> >                             () -> startNewEntityTaskLane(adTask, transportService),
> >                             TimeValue.timeValueSeconds(SLEEP_TIME_FOR_NEXT_ENTITY_TASK_IN_MILLIS),
> >                             AD_BATCH_TASK_THREAD_POOL_NAME
> >                         );
> >                 }
> >             }
> > ```
> 
> Good catch, will fix the comment. Let me know when you finish reviewing all files. Will fix all comments in a new commit.

I finished reviewing all files.

ylwu-amzn commented on src/main/java/org/opensearch/ad/task/ADTaskCacheManager.java:
```diff
@@ -872,9 +928,9 @@ public void addPendingEntity(String detectorId, String entity) {
-    public synchronized void moveToRunningEntity(String detectorId, String entity) {
-        if (this.hcBatchTaskCaches.containsKey(detectorId)) {
-            ADHCBatchTaskCache hcTaskCache = this.hcBatchTaskCaches.get(detectorId);
+    public void moveToRunningEntity(String detectorId, String entity) {
```
One entity can only be ran by one thread, no need to add synchronized here.

ylwu-amzn commented on src/main/java/org/opensearch/ad/task/ADHCBatchTaskCache.java:
```diff
@@ -289,19 +266,13 @@ public int increaseTaskRetry(String taskId) {
-            runningEntities.remove(entity);
+            return false;
         }
+        return tempEntities.remove(entity) || pendingEntities.remove(entity) || runningEntities.remove(entity);
```
Entity can only exist in one queue. So should not need to delete from other queues if it removed from one queue first. 
Maybe add more checking is good with some minor performance loss by checking extra queues.
Will remove from all queues first, then return. The cons is if entity only exists in temp queue, we need to do extra checking in pending and running queue.

ylwu-amzn commented on src/main/java/org/opensearch/ad/transport/CronTransportAction.java:
```diff
@@ -112,7 +112,7 @@ protected CronNodeResponse newNodeResponse(StreamInput in) throws IOException {
     @Override
     protected CronNodeResponse nodeOperation(CronNodeRequest request) {
-        LOG.debug("Start running AD hourly cron. ");
+        LOG.info("Start running AD hourly cron.");
```
From performance test, I see hourly spikes on metrics. I guess they are caused by hourly cron, but we didn't print info log so I don't know when the hourly cron job starts. Add this info level log could benefit such root cause analysis. The cons is we have one more line of info log every hour. I think that's acceptable.

ylwu-amzn commented on src/main/java/org/opensearch/ad/task/ADTaskManager.java:
```diff
@@ -605,54 +598,28 @@ public void checkTaskSlots(
+                // We may assign more task slots. For example, cluster has 4 data nodes, each node can run 2
+                // batch tasks, so the available task slot number is 8. If max running entities per HC is 4,
+                // then we will assign 4 tasks slots to this HC detector (4 is less than 8). The data index
+                // only have 2 entities. So we assign 2 more task slots than actual need. But it's ok as we
```
sure, will fix

ylwu-amzn commented on src/main/java/org/opensearch/ad/task/ADTaskCacheManager.java:
```diff
@@ -1132,44 +1212,175 @@ public String pollDeletedDetector() {
+                String detectorId = detectorRunStates.getKey();
+                boolean noRunningTask = isNullOrEmpty(getTasksOfDetector(detectorId));
+                Map<String, ADHCBatchTaskRunState> taskRunStates = detectorRunStates.getValue();
+                if (taskRunStates == null || !noRunningTask) {
```
Good catch, will fix

ohltyler commented:
@ylwu-amzn when merged, I'll add this to the list of PRs to backport to `1.x`/`1.1`.
